docker/lr-pb/shard_bam.py

```diff
@@ -0,0 +1,124 @@
+import argparse
+import gzip
+from collections import OrderedDict
+from math import ceil
+import pysam
+from construct import *
+
+from multiprocessing.pool import ThreadPool
+from functools import partial
+
+
+def write_shard(bam, offsets, prefix, index):
+    """
+    Write subsection of bam to a shard.
+    """
+
+    bf = pysam.Samfile(bam, 'rb', check_sq=False)
+
+    # Advance to the specified virtual file offset.
+    bf.seek(offsets[index])
+
+    num_reads = 0
+    with pysam.Samfile(f'{prefix}{index}.bam', 'wb', header=bf.header) as out:
+        # Write until we've advanced to (but haven't written) the read that begins the next shard.
+        while True:
+            read = bf.__next__()
+            out.write(read)
+
+            num_reads += 1
+            if bf.tell() >= offsets[index+1]:
+                break
+
+    return num_reads
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description='Shard .bam file using the .pbi index', prog='shard_bam')
+    parser.add_argument('-p', '--prefix', type=str, default="shard", help="Shard filename prefix")
+    parser.add_argument('-n', '--num_shards', type=int, default=4, help="Number of shards")
+    parser.add_argument('-t', '--num_threads', type=int, default=2, help="Number of threads to use during sharding")
+    parser.add_argument('-i', '--index', type=str, required=False, help="Index filename")
+    parser.add_argument('bam', type=str, help="BAM")
+    args = parser.parse_args()
+
+    pbi = args.bam + ".pbi" if args.index is None else args.index
+
+    print(f"Reading index ({pbi}). This may take a few minutes...", flush=True)
+
+    # Decode PacBio .pbi file.  This is not a full decode of the index, only the parts we need for sharding.
+    # More on index format at https://pacbiofileformats.readthedocs.io/en/9.0/PacBioBamIndex.html .
+
+    fmt = Struct(
```

Comment:
I believe lines from here and on should be re-factored, given this step is of critical importance (if shard-ed wrong, all downstream works are unreliable)

Reply:
ping again

Reply:
I've refactored this into a new function ("compute_shard_offsets") that returns a list of all valid shard offsets and the frequency for all ZMW hole numbers.

The "write_shard" function now accepts these shard offsets and the ZMW hole number frequencies.  This method checks to see if all reads from a given ZMW have actually been written to this shard (line 41-42).  This means that if a file is somehow missorted and reads from the same ZMW occur out of order, the assertion will be tripped and the script will deliberately halt in failure.

Reply:
This new refactor is much better!